### ISSUE TITLE
Incorrect event-reduce action

### ISSUE THREAD

**Issue by TristanH** [pubkey/event-reduce#444]
We discovered some scary behavior in our app where items were being sorted wrong. We dug into and realized that event-reduce is actually suggesting an `insertLast` when it should be an `insertFirst`.

Here's the test case showing it: https://github.com/pubkey/event-reduce/pull/443/

I thiiink (though I don't fully understand the library yet) the issue is somewhere in the truth table to bdd conversion?

I think this because the json truth table `calculateActionFromMap` seems to correctly tell us `runFullQueryAgain` (well it's not ideal, which would be `insertFirst`, but it doesn't give us incorrect results) for this case of 0101000000010011001.

However, `calculateActionName` is wrong. It tells us to `insertLast`.

-- 

We can obviously patch this, but the issue opens up some serious questions for us:

1. From reading the iterative fuzzing process, it seems there is no guarantee of correctness -- is that true?
2. How can we be confident there are no further issues like this?

**Comment by pubkey** [comment]
Hi @TristanH 
I reproduced the error and I am working on a fix.

The state calculation is correct:
```
> logStateSet(getStateSet(input));

state: isInsert : 0
state: isUpdate : 1
state: isDelete : 0
state: hasLimit : 1
state: isFindOne : 0
state: hasSkip : 0
state: wasResultsEmpty : 0
state: previousUnknown : 0
state: wasLimitReached : 0
state: wasFirst : 0
state: wasLast : 0
state: sortParamsChanged : 1
state: wasInResult : 0
state: wasSortedBeforeFirst : 0
state: wasSortedAfterLast : 1
state: isSortedBeforeFirst : 1
state: isSortedAfterLast : 0
state: wasMatching : 0
state: doesMatchNow : 1
```

I am now working on a refactor of the whole module to bring things up to date. Mostly the stripping stuff we no longer need (`previousUnknown`) because RxDB no longer runs on PouchDB.


> 1. From reading the iterative fuzzing process, it seems there is no guarantee of correctness -- is that true?

There are 2^13 (=8192) possible state combinations, the goal is to ensure every possibility is touched at least once. I do not see a reason why this should not be possible. In the refactoring I will remove the `previousUnknown` state, so we only have 2^12 (=4096) states combinations.

> 2. How can we be confident there are no further issues like this?

I am not sure if it can ever be proven to be 100% correct, but that is also true for all code you use.
My best option is to add more tests and improve the fuzzing performance and runtime.

**Comment by TristanH** [comment]
Hi @pubkey, that's great news thank you!

> I am now working on a refactor of the whole module to bring things up to date. Mostly the stripping stuff we no longer need (previousUnknown) because RxDB no longer runs on PouchDB.

Aha great yeah, I was wondering about that since I never saw it used.

> EDIT: We have 19 state functions, which makes 2^19 (=524288) state combinations. In the current refactoring I am removing one of them but it is still 262144 combinations. I do not see that as a big problem. Last time I let the fuzzing run on a single core for one month. I will make it use more CPUs to tackle the problem.

Awesome!! Yeah the single threadedness when I ran the fuzzing did seem to be a bottleneck. We're also happy to donate some compute (eg the biggest digitalocean cpu-optimized droplet or something) to run this if you like.

> I am not sure if it can ever be proven to be 100% correct, but that is also true for all code you use.
My best option is to add more tests and improve the fuzzing performance and runtime.
If this seems to risky for your use case, you should disable eventReduce in RxDB in the meantime.

For sure, I didn't meant to set an impossibly high standard. The benefits of event-reduce are already super valuable to us. To be specific, I was wondering if there was a way to guarantee that the fuzzing results don't give an incorrect action like `insertLast`, but instead return `runFullQuery`. But I guess your other comment about using more cores (plus more tests) should resolve it.

Thanks again!

**Comment by TristanH** [comment]
@pubkey in case it's useful, we found another broken case:

```
isInsert false
isUpdate true
isDelete false
hasLimit true
isFindOne false
hasSkip false
wasResultsEmpty false
previousUnknown false
wasLimitReached true
wasFirst false
wasLast false
sortParamsChanged true
wasInResult false
wasSortedBeforeFirst false
wasSortedAfterLast true
isSortedBeforeFirst true
isSortedAfterLast false
wasMatching false
doesMatchNow true

Expected: removeLastInsertFirst
Actual: doNothing
```

**Comment by pubkey** [comment]
Hi @TristanH can you also add this as a test so that I can see how you reached that state set?

Current state of my debugging:
So the root issue here is that the fuzzing never reached the given state set `0101000000010011001`. This must be the case because the state set is not part of the truth-table.json Your test case is pretty simple so there must be a bug in the fuzzing test creator. I will update you again on process.

**Comment by TristanH** [comment]
> States that are not known to the truth-table will then produce an unknown result in the binary-decision-diagram, because the bdd does not care about unknown states and will return whatever state-chain the optimization has created.

Ahhh, gotcha! That explains it.

**Comment by TristanH** [comment]
@pubkey do you have any test procedures that modify two fields in the same update? that's what happens in all 3 broken cases, so wondering if that's the potential missing case.

**Comment by pubkey** [comment]
@TristanH  that is exactly what I was thinking about a few minutes ago.

So the main changes atm are:

- Better random query generation: https://github.com/pubkey/event-reduce/pull/445/commits/a3c7002e2c68080c0c80ee4b5df2fa2d50e9f169#diff-e3b9537b24b427eebaa0e38f8a980b22234dd1ba20a1a65092d35d35f2b18485R182
- Better random event generation: https://github.com/pubkey/event-reduce/pull/445/commits/9331c1629be65501fcdf80883b375897a9fe42e1#diff-a7a98b6c47fa60580c301fabb7647b20bc1c7b6b08e23d68effd419ac4732e8bR53

I will now change up to all 3 fields or even no field at all.

**Comment by TristanH** [comment]
@pubkey that's great news! Yes, the improvement to randomChangeHuman looks super promising :)

I pushed up two more test cases (all the same update, but different queries) to https://github.com/pubkey/event-reduce/pull/443

**Comment by pubkey** [comment]
@TristanH 

I tried to also add your tests, I have a question:
The test `new first item with more than limit results`  expects `insertFirst` as action but atm it gets `removeLastInsertFirst`. It think `removeLastInsertFirst` is correct because the query results before already "filled" the limit of 2 results. So using `insertFirst` would give a wrong result set of 3 documents. Isn't `removeLastInsertFirst` correct?


Atm the test `new item should push all results in skip query down` fails on my new version which is correct because the fuzzing did not reach that state so far. But shouldn't the action there also be `removeLastInsertFirst` instead of `insertFirst`?

I am running the fuzzing in parallel on 8 cores now, maybe the third test will be solve by tomorrow.

**Comment by TristanH** [comment]
@pubkey correct I think!

`new first item with fewer than limit results` should be `insertFirst`

'new first item with more than limit results' should be `removeLastInsertFirst`

'new item should push all results in skip query down' should be `runFullQueryAgain`. Because of the skip pushing a new item into the results, it's impossible to be cached in theory. Gah, sorry about my comment there -- i copied that over incorrectly.

> I am running the fuzzing in parallel on 8 cores now, maybe the third test will be solved by tomorrow.

That's great news. We have a digitalocean box with 48 intel cores if you wanna use that, we can give you ssh access if you want to run there on ubuntu.

<img width="216" alt="CleanShot 2023-11-23 at 18 44 16@2x" src="https://github.com/pubkey/event-reduce/assets/1284332/e46ac576-bfbc-46ca-9eb2-cee69e8b1821">

**Comment by pubkey** [comment]
> That's great news. We have a digitalocean box with 48 intel cores if you wanna use that, we can give you ssh access if you want to run there on ubuntu.

Thank you for the offering. I am using my own server atm. Last time I ran the fuzzing it did not find any new states after 2 weeks. So having it run in parallel 8x I think it should be fine. Also I switched out minimongo with the mingo package, so fuzzing runs faster by itself now.
My plan is to let it run for a few months even after everthing is fixed and the new version is released.

**Comment by TristanH** [comment]
> Thank you for the offering. I will be using my own server atm. Last time I ran the fuzzing it did not find any new states after 2 weeks. So having it run in parallel 8x I think it should be fine. Also I switched out minimongo with the mingo package, so fuzzing runs faster by itself now.
My plan is to let it run for a few months even after everthing is fixed and the new version is released.

Great! sounds good to me, looking forward to the new version :)

**Comment by pubkey** [comment]
After 1d of fuzzing, it found out by itself that your 3rd test needs the action 'runFullQueryAgain' :partying_face: 
I will let it run over the weekend and then release a new version of the RxDB beta with the updated event-reduce.
Until RxDB v15 is out of beta, the fuzzing can run for some weeks.

**Comment by TristanH** [comment]
@pubkey great! as soon as you release we will also test in our app.

**Comment by TristanH** [comment]
@pubkey i saw that 5.0.0 merged -- is that ready for us to test? i will point our fork of rxdb to use it if so :)

**Comment by TristanH** [comment]
@pubkey ok great, thanks!

**Comment by TristanH** [comment]
@pubkey that's great news! We'll get on with testing it.

And thanks for the explanation! That all makes sense, and I do feel much more confident in the library now. Thank you for your quick and detailed work :)

(I am happy to close this issue now if you want)
